File: src/ui/pages/category.py
import json
import threading
from gi.repository import Adw, GLib, GObject, Gtk, Pango
from api.client import MusicClient
from ui.context_menu import MenuAction, show_item_menu, show_song_menu
from ui.util_classes import ScrolledWindow
from ui.utils import AsyncImage, AsyncPicture, LikeButton, copy_to_clipboard, parse_item_metadata
from ui.widgets.scroll_box import HorizontalScrollBox
from ui.widgets.media_card import MediaCardWidget


class CategoryPage(Adw.Bin):
    __gsignals__ = {
        "header-title-changed": (GObject.SignalFlags.RUN_FIRST, None, (str,))
    }

    def __init__(self, player, open_playlist_callback, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.player = player
        self.open_playlist_callback = open_playlist_callback
        self.client = MusicClient()
        self.params = None
        self.title = ""
        self._is_loading = False
        self._section_limits = {}
        self._cached_sections = None
        self._cached_params = None

        # Main Layout
        self.main_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)

        # Scrolled Window
        self.scrolled = ScrolledWindow()
        self.scrolled.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.AUTOMATIC)
        self.scrolled.set_vexpand(True)

        vadjust = self.scrolled.get_vadjustment()
        vadjust.connect("value-changed", self._on_scroll)

        # Content Box
        self.content_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=32)
        self.content_box.set_margin_top(24)
        self.content_box.set_margin_bottom(24)
        self.content_box.set_margin_start(16)
        self.content_box.set_margin_end(16)

        # Title Label
        self.page_title_label = Gtk.Label(label="")
        self.page_title_label.add_css_class("title-1")
        self.page_title_label.set_halign(Gtk.Align.START)
        self.page_title_label.set_margin_bottom(16)
        self.content_box.append(self.page_title_label)

        # Loading Spinner
        self._loading_wrap = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
        self._loading_wrap.set_vexpand(True)
        self._loading_wrap.set_valign(Gtk.Align.CENTER)
        self._loading_wrap.set_halign(Gtk.Align.CENTER)
        self._loading_wrap.set_margin_top(32)
        self._loading_wrap.set_margin_bottom(32)
        self.loading_spinner = Adw.Spinner()
        self.loading_spinner.set_halign(Gtk.Align.CENTER)
        self.loading_spinner.set_size_request(48, 48)
        self._loading_wrap.append(self.loading_spinner)
        self._loading_wrap.set_visible(False)
        self.content_box.append(self._loading_wrap)

        # Clamp for consistent width
        self.clamp = Adw.Clamp()
        self.clamp.set_maximum_size(1024)
        self.clamp.set_tightening_threshold(600)
        self.clamp.set_child(self.content_box)

        self.scrolled.set_child(self.clamp)
        self.main_box.append(self.scrolled)

        self.set_child(self.main_box)

    def _on_scroll(self, vadjust):
        if vadjust.get_value() > 50:
            self.emit("header-title-changed", self.title)
        else:
            self.emit("header-title-changed", "")

    def set_compact_mode(self, compact):
        self._compact = compact
        self._propagate_compact(self.content_box, compact)

        if compact:
            self.add_css_class("compact")
            self.content_box.set_spacing(16)
        else:
            self.remove_css_class("compact")
            self.content_box.set_spacing(32)

    def _propagate_compact(self, widget, compact):
        if hasattr(widget, "set_compact") and hasattr(widget, "target_size"):
            widget.set_compact(compact)
        child = widget.get_first_child() if hasattr(widget, "get_first_child") else None
        while child:
            self._propagate_compact(child, compact)
            child = child.get_next_sibling()

    def load_category(self, params, title):
        self.params = params
        self.title = title
        self.page_title_label.set_label(title)

        child = self.content_box.get_first_child()
        while child:
            next_child = child.get_next_sibling()
            if child != self._loading_wrap and child != self.page_title_label:
                self.content_box.remove(child)
            child = next_child

        self.emit("header-title-changed", title)
        self._load_data()

    def _load_data(self):
        if self._is_loading:
            return

        self._is_loading = True
        self._loading_wrap.set_visible(True)

        if self._cached_sections is not None and self._cached_params == self.params:
            GLib.idle_add(self._render_sections, self._cached_sections)
            return

        def fetch_func():
            try:
                sections = self.client.get_category_page(self.params)
                self._cached_sections = sections
                self._cached_params = self.params
                GLib.idle_add(self._render_sections, sections)
            except Exception as e:
                print(f"Error loading category page: {e}")
                GLib.idle_add(lambda: self._loading_wrap.set_visible(False))
                self._is_loading = False

        threading.Thread(target=fetch_func, daemon=True).start()

    def _render_sections(self, sections):
        child = self.content_box.get_first_child()
        while child:
            next_child = child.get_next_sibling()
            if child != self._loading_wrap and child != self.page_title_label:
                self.content_box.remove(child)
            child = next_child

        if sections:
            for section in sections:
                is_video_section = "video" in section["title"].lower()
                is_song_section = section["title"].lower() == "songs" or (
                    not is_video_section
                    and all(not i.get("browseId") and i.get("videoId") for i in section["items"][:3])
                )

                if is_song_section:
                    self._add_songs_list(section["title"], section["items"])
                else:
                    self._add_carousel(section["title"], section["items"])

        self._is_loading = False
        self._loading_wrap.set_visible(False)

    def _make_card(self, item):
        card = MediaCardWidget(
            item,
            player=self.player,
            title_lines=1,
            on_clicked=lambda btn, it: self._on_item_clicked(None, 1, 0, 0, it)
        )
    
        gesture = Gtk.GestureClick()
        gesture.set_button(3)
        gesture.connect("released", self.on_grid_right_click, card)
        card.add_controller(gesture)
    
        lp = Gtk.GestureLongPress()
        lp.connect(
            "pressed",
            lambda g, x, y, c=card: self.on_grid_right_click(g, 1, x, y, c),
        )
        card.add_controller(lp)
        return card

    def _add_carousel(self, title, items):
        if not items:
            return

        section_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=16)
        self.content_box.append(section_box)

        label = Gtk.Label(label=title)
        label.add_css_class("heading")
        label.set_halign(Gtk.Align.START)
        section_box.append(label)

        scroll_box = HorizontalScrollBox()
        inner_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=16)

        for item in items:
            card = self._make_card(item)
            inner_box.append(card)

        scroll_box.set_content(inner_box)
        section_box.append(scroll_box)

    def _add_songs_list(self, title, items):
        if not items:
            return

        section_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=8)
        self.content_box.append(section_box)

        label = Gtk.Label(label=title)
        label.add_css_class("heading")
        label.set_halign(Gtk.Align.START)
        label.set_margin_bottom(8)
        section_box.append(label)

        list_box = Gtk.ListBox()
        list_box.add_css_class("boxed-list")
        list_box.add_css_class("songs-list")
        list_box.set_selection_mode(Gtk.SelectionMode.NONE)

        limit = self._section_limits.get(title, 5)
        showing_items = items[:limit]

        for item in showing_items:
            row = Gtk.ListBoxRow()
            box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=12)
            box.add_css_class("song-row")
            row.set_child(box)

            thumbnails = item.get("thumbnails", [])
            thumb_url = thumbnails[-1]["url"] if thumbnails else None

            img = AsyncPicture(
                url=thumb_url,
                target_size=56,
                crop_to_square=True,
                player=self.player,
            )
            img.video_id = item.get("videoId")
            img.add_css_class("song-img")
            root = self.get_root()
            img.set_compact(getattr(root, "_is_compact", False) if root else False)
            box.append(img)

            song_title = item.get("title", "Unknown")

            artist_list = item.get("artists", [])
            subtitle = ""
            if isinstance(artist_list, list):
                subtitle = ", ".join(
                    [
                        a.get("name", "Unknown")
                        for a in artist_list
                        if isinstance(a, dict)
                    ]
                )
            else:
                subtitle = artist_list or ""

            vbox = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=2)
            vbox.set_valign(Gtk.Align.CENTER)
            vbox.set_hexpand(True)

            title_label = Gtk.Label(label=song_title)
            title_label.set_halign(Gtk.Align.START)
            title_label.set_ellipsize(Pango.EllipsizeMode.END)
            title_label.set_lines(1)
            title_label.set_width_chars(1)
            title_label.set_xalign(0.0)

            subtitle_label = Gtk.Label(label=subtitle)
            subtitle_label.set_halign(Gtk.Align.START)
            subtitle_label.set_ellipsize(Pango.EllipsizeMode.END)
            subtitle_label.set_lines(1)
            subtitle_label.set_width_chars(1)
            subtitle_label.set_xalign(0.0)
            subtitle_label.add_css_class("dim-label")
            subtitle_label.add_css_class("caption")

            title_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=6)
            title_box.append(title_label)

            meta = parse_item_metadata(item)
            if meta["is_explicit"]:
                explicit_badge = Gtk.Label(label="E")
                explicit_badge.add_css_class("explicit-badge")
                explicit_badge.set_valign(Gtk.Align.CENTER)
                title_box.append(explicit_badge)

            vbox.append(title_box)
            vbox.append(subtitle_label)
            box.append(vbox)

            if item.get("videoId"):
                like_btn = LikeButton(
                    self.client, item["videoId"], item.get("likeStatus", "INDIFFERENT")
                )
                like_btn.set_valign(Gtk.Align.CENTER)
                box.append(like_btn)

            row.item_data = item
            list_box.append(row)

            click_gesture = Gtk.GestureClick()
            click_gesture.set_button(1)
            click_gesture.connect("released", self._on_item_clicked, item)
            row.add_controller(click_gesture)

            right_click = Gtk.GestureClick()
            right_click.set_button(3)
            right_click.connect("released", self.on_song_right_click, row)
            row.add_controller(right_click)

        section_box.append(list_box)

        if len(items) > limit:
            show_all_btn = Gtk.Button(label="View All")
            show_all_btn.add_css_class("pill")
            show_all_btn.set_halign(Gtk.Align.CENTER)
            show_all_btn.set_margin_top(12)

            btn_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=8)
            btn_box.set_halign(Gtk.Align.CENTER)
            btn_box.append(show_all_btn)

            show_all_btn.connect("clicked", lambda btn, t=title: self.on_show_all_songs_clicked(t))
            section_box.append(btn_box)

    def on_show_all_songs_clicked(self, title):
        self._section_limits[title] = 1000
        if self._cached_sections:
            self._render_sections(self._cached_sections)

    def on_grid_right_click(self, gesture, n_press, x, y, item_box):
        if not hasattr(item_box, "item_data"):
            return
        data = item_box.item_data
        show_item_menu(
            item_box,
            x,
            y,
            data,
            player=self.player,
            client=self.client,
            prefix="item",
            extras=[
                MenuAction(
                    "Copy JSON (Debug)",
                    lambda d=data: copy_to_clipboard(json.dumps(d, indent=2)),
                    section="debug",
                )
            ],
        )

    def on_song_right_click(self, gesture, n_press, x, y, row):
        if not hasattr(row, "item_data"):
            return
        show_song_menu(
            row,
            x,
            y,
            row.item_data,
            player=self.player,
            client=self.client,
            prefix="row",
        )

    def _on_item_clicked(self, gesture, n_press, x, y, item):
        video_id = item.get("videoId")
        browse_id = item.get("browseId") or item.get("playlistId")

        if video_id:
            self.player.play_tracks([item])
        elif browse_id:
            initial_data = {
                "title": item.get("title", ""),
                "thumb": (item.get("thumbnails", [{}])[-1] or {}).get("url")
                if item.get("thumbnails")
                else None,
            }
            try:
                self.open_playlist_callback(browse_id, initial_data)
            except TypeError:
                self.open_playlist_callback(browse_id)

File: rust/src/ui/pages/category.rs
//! Port of ui/pages/category.py: the carousels behind one mood or genre pill.
//! Song shelves become boxed lists that show five rows until View All opens
//! the rest; everything else becomes a card strip.

use std::cell::{Cell, RefCell};
use std::collections::HashMap;
use std::rc::Rc;

use gtk::{glib, prelude::*};

use crate::model::MediaItem;
use crate::net::explore::{self, CategorySection};
use crate::ui::context::UiContext;
use crate::ui::context_menu::{MenuAction, Section, show_item_menu_with};
use crate::ui::copy_to_clipboard;
use crate::ui::pages::activate_item;
use crate::ui::widgets::media_card::{CardOptions, MediaCard, STRIP_SPACING, STRIP_SPACING_COMPACT};
use crate::ui::widgets::scroll_box::HorizontalScrollBox;
use crate::ui::widgets::song_row::SongRow;

/// Rows a song shelf shows before View All.
const SONG_PREVIEW: usize = 5;
/// What View All raises that limit to.
const SONG_ALL: usize = 1000;
/// How far the page scrolls before the header takes over the title.
const TITLE_HANDOVER: f64 = 50.0;

type TitleListener = Box<dyn Fn(&str)>;

pub struct CategoryPage {
    root: gtk::Box,
    content_box: gtk::Box,
    title_label: gtk::Label,
    loading_wrap: gtk::Box,
    ctx: Rc<UiContext>,
    params: RefCell<String>,
    title: RefCell<String>,
    is_loading: Cell<bool>,
    /// How many rows each song shelf shows, raised by that shelf's View All.
    section_limits: RefCell<HashMap<String, usize>>,
    sections: RefCell<Vec<CategorySection>>,
    cards: RefCell<Vec<Rc<MediaCard>>>,
    strips: RefCell<Vec<gtk::Box>>,
    scrollers: RefCell<Vec<Rc<HorizontalScrollBox>>>,
    rows: RefCell<Vec<Rc<SongRow>>>,
    on_title: RefCell<Option<TitleListener>>,
}

impl CategoryPage {
    pub fn new(ctx: Rc<UiContext>) -> Rc<Self> {
        let root = gtk::Box::builder().orientation(gtk::Orientation::Vertical).build();
        let scrolled = gtk::ScrolledWindow::builder().hscrollbar_policy(gtk::PolicyType::Never).vscrollbar_policy(gtk::PolicyType::Automatic).vexpand(true).build();
        crate::ui::suppress_hover_while_scrolling(&scrolled);

        let content_box = gtk::Box::builder().orientation(gtk::Orientation::Vertical).spacing(32).margin_top(24).margin_bottom(24).margin_start(16).margin_end(16).build();
        let title_label = gtk::Label::builder().css_classes(["title-1"]).halign(gtk::Align::Start).margin_bottom(16).build();
        content_box.append(&title_label);

        let loading_wrap = gtk::Box::builder().orientation(gtk::Orientation::Vertical).vexpand(true).valign(gtk::Align::Center).halign(gtk::Align::Center).margin_top(32).margin_bottom(32).visible(false).build();
        let spinner = adw::Spinner::builder().halign(gtk::Align::Center).build();
        spinner.set_size_request(48, 48);
        loading_wrap.append(&spinner);
        content_box.append(&loading_wrap);

        let clamp = adw::Clamp::builder().maximum_size(1024).tightening_threshold(600).child(&content_box).build();
        scrolled.set_child(Some(&clamp));
        root.append(&scrolled);

        let page = Rc::new(Self {
            root,
            content_box,
            title_label,
            loading_wrap,
            ctx,
            params: RefCell::new(String::new()),
            title: RefCell::new(String::new()),
            is_loading: Cell::new(false),
            section_limits: RefCell::new(HashMap::new()),
            sections: RefCell::new(Vec::new()),
            cards: RefCell::new(Vec::new()),
            strips: RefCell::new(Vec::new()),
            scrollers: RefCell::new(Vec::new()),
            rows: RefCell::new(Vec::new()),
            on_title: RefCell::new(None),
        });

        // Past the first scroll the page title moves into the header bar.
        let weak = Rc::downgrade(&page);
        scrolled.vadjustment().connect_value_changed(move |adj| {
            let Some(p) = weak.upgrade() else { return };
            let title = p.title.borrow().clone();
            p.emit_title(if adj.value() > TITLE_HANDOVER { &title } else { "" });
        });
        let weak = Rc::downgrade(&page);
        page.ctx.on_compact(move |compact| match weak.upgrade() {
            Some(p) => {
                p.set_compact_mode(compact);
                true
            }
            None => false,
        });
        page
    }

    pub fn widget(&self) -> &gtk::Box {
        &self.root
    }

    /// The window sets its title from this, like the header-title-changed signal.
    pub fn set_on_header_title(&self, f: impl Fn(&str) + 'static) {
        self.on_title.replace(Some(Box::new(f)));
    }

    fn emit_title(&self, title: &str) {
        if let Some(f) = self.on_title.borrow().as_ref() {
            f(title);
        }
    }

    pub fn set_compact_mode(&self, compact: bool) {
        if compact {
            self.root.add_css_class("compact");
            self.content_box.set_spacing(16);
        } else {
            self.root.remove_css_class("compact");
            self.content_box.set_spacing(32);
        }
        for strip in self.strips.borrow().iter() {
            strip.set_spacing(if compact { STRIP_SPACING_COMPACT } else { STRIP_SPACING });
        }
        for card in self.cards.borrow().iter() {
            card.set_compact(compact);
        }
    }

    pub fn load_category(self: &Rc<Self>, params: &str, title: &str) {
        self.params.replace(params.to_owned());
        self.title.replace(title.to_owned());
        self.title_label.set_label(title);
        self.section_limits.borrow_mut().clear();
        self.clear_sections();
        self.emit_title(title);

        if self.is_loading.replace(true) {
            return;
        }
        self.loading_wrap.set_visible(true);
        let api = self.ctx.net.client().api();
        let params = params.to_owned();
        let handle = self.ctx.net.spawn(async move { explore::get_category_page(&api, &params).await });
        let weak = Rc::downgrade(self);
        glib::spawn_future_local(async move {
            let outcome = handle.await;
            let Some(page) = weak.upgrade() else { return };
            page.is_loading.set(false);
            page.loading_wrap.set_visible(false);
            match outcome {
                Ok(Ok(sections)) => {
                    page.sections.replace(sections);
                    page.render_sections();
                }
                Ok(Err(err)) => tracing::warn!(%err, "category page failed"),
                Err(_) => {}
            }
        });
    }

    /// Everything but the title and the spinner, which the page keeps.
    fn clear_sections(&self) {
        let mut child = self.content_box.first_child();
        while let Some(widget) = child {
            child = widget.next_sibling();
            if widget != self.title_label.clone().upcast::<gtk::Widget>() && widget != self.loading_wrap.clone().upcast::<gtk::Widget>() {
                self.content_box.remove(&widget);
            }
        }
        self.cards.borrow_mut().clear();
        self.strips.borrow_mut().clear();
        self.scrollers.borrow_mut().clear();
        self.rows.borrow_mut().clear();
    }

    fn render_sections(self: &Rc<Self>) {
        self.clear_sections();
        for section in self.sections.borrow().iter() {
            if is_song_section(section) {
                self.add_songs_list(&section.title, &section.items);
            } else {
                self.add_carousel(&section.title, &section.items);
            }
        }
        self.set_compact_mode(self.ctx.compact.get());
    }

    fn add_carousel(self: &Rc<Self>, title: &str, items: &[MediaItem]) {
        if items.is_empty() {
            return;
        }
        let section = gtk::Box::builder().orientation(gtk::Orientation::Vertical).spacing(16).build();
        section.append(&heading(title));
        let scroll_box = HorizontalScrollBox::new();
        let compact = self.ctx.compact.get();
        let strip = gtk::Box::builder().orientation(gtk::Orientation::Horizontal).spacing(if compact { STRIP_SPACING_COMPACT } else { STRIP_SPACING }).build();
        let pool: Vec<MediaItem> = items.iter().filter(|i| i.kind.is_playable()).cloned().collect();
        for item in items {
            let card = MediaCard::new(&self.ctx, item.clone(), CardOptions { title_lines: 1, ..CardOptions::default() });
            let ctx = self.ctx.clone();
            let pool_c = pool.clone();
            card.connect_clicked(move |item| activate_item(&ctx, item, &pool_c));
            self.attach_card_menu(card.widget(), item);
            strip.append(card.widget());
            self.cards.borrow_mut().push(card);
        }
        scroll_box.set_content(&strip);
        section.append(scroll_box.widget());
        self.content_box.append(&section);
        self.strips.borrow_mut().push(strip);
        self.scrollers.borrow_mut().push(scroll_box);
    }

    fn add_songs_list(self: &Rc<Self>, title: &str, items: &[MediaItem]) {
        if items.is_empty() {
            return;
        }
        let section = gtk::Box::builder().orientation(gtk::Orientation::Vertical).spacing(8).build();
        let label = heading(title);
        label.set_margin_bottom(8);
        section.append(&label);

        let limit = self.section_limits.borrow().get(title).copied().unwrap_or(SONG_PREVIEW);
        let showing = &items[..items.len().min(limit)];
        let list = gtk::ListBox::builder().css_classes(["boxed-list", "songs-list"]).selection_mode(gtk::SelectionMode::None).build();
        let pool: Vec<MediaItem> = showing.iter().filter(|i| i.kind.is_playable()).cloned().collect();
        for item in showing {
            let row = SongRow::new(self.ctx.clone());
            row.set_search_style(true);
            row.set_plain_subtitle(Some(item.artists_text()));
            row.bind(item, None);
            list.append(row.widget());
            self.rows.borrow_mut().push(row);
        }
        let ctx = self.ctx.clone();
        let shown = showing.to_vec();
        list.connect_row_activated(move |_, row| {
            if let Some(item) = shown.get(row.index().max(0) as usize) {
                activate_item(&ctx, item, &pool);
            }
        });
        section.append(&list);

        if items.len() > limit {
            let button = gtk::Button::builder().label("View All").css_classes(["pill"]).halign(gtk::Align::Center).margin_top(12).build();
            let weak = Rc::downgrade(self);
            let title = title.to_owned();
            button.connect_clicked(move |_| {
                if let Some(page) = weak.upgrade() {
                    page.section_limits.borrow_mut().insert(title.clone(), SONG_ALL);
                    page.render_sections();
                }
            });
            let button_box = gtk::Box::builder().orientation(gtk::Orientation::Horizontal).spacing(8).halign(gtk::Align::Center).build();
            button_box.append(&button);
            section.append(&button_box);
        }
        self.content_box.append(&section);
    }

    /// Card menu with the Copy JSON entry on_grid_right_click added.
    fn attach_card_menu(self: &Rc<Self>, widget: &gtk::Button, item: &MediaItem) {
        let ctx = self.ctx.clone();
        let button = widget.clone();
        let item = item.clone();
        let open = Rc::new(move |x: f64, y: f64| {
            let data = item.clone();
            let extras = vec![MenuAction::new("Copy JSON (Debug)", Section::Clipboard, move || {
                if let Ok(text) = serde_json::to_string_pretty(&data) {
                    copy_to_clipboard(&text);
                }
            })];
            show_item_menu_with(&button, x, y, &item, &ctx, extras);
        });
        let right = gtk::GestureClick::builder().button(gtk::gdk::BUTTON_SECONDARY).build();
        let o = open.clone();
        right.connect_released(move |_, _, x, y| o(x, y));
        widget.add_controller(right);
        let long = gtk::GestureLongPress::new();
        long.connect_pressed(move |_, x, y| open(x, y));
        widget.add_controller(long);
    }
}

/// Port of the shelf test in _render_sections: a shelf named Songs, or one
/// whose first rows are all playable and whose name says nothing about video.
fn is_song_section(section: &CategorySection) -> bool {
    let title = section.title.to_lowercase();
    if title == "songs" {
        return true;
    }
    !title.contains("video") && section.items.iter().take(3).all(|item| item.kind.is_playable())
}

fn heading(title: &str) -> gtk::Label {
    gtk::Label::builder().label(title).css_classes(["heading"]).halign(gtk::Align::Start).build()
}
